### PR TITLE
Migrate `snforge_std` to edition `2024_07`

### LINKED ISSUE #2638: Migrate `snforge_std` to edition `2024_07`

### PR DESCRIPTION
**Pull request by kkawula**
Closes #2638

## Checklist

- [x] Linked relevant issue
- [x] Updated relevant documentation
- [x] Added relevant tests
- [x] Performed self-review of the code
- [x] Added changes to `CHANGELOG.md`

### CODE REVIEW

File: snforge_std/src/cheatcodes.cairo [hunk 1]
```diff
@@ -1,20 +1,20 @@
 use starknet::{ContractAddress, ClassHash, contract_address_const};
 use starknet::testing::cheatcode;
 use super::_cheatcode::handle_cheatcode;
-use execution_info::{cheat_execution_info, ExecutionInfoMock, CheatArguments, Operation};
+pub use execution_info::{cheat_execution_info, ExecutionInfoMock, CheatArguments, Operation};
```
Comment (cptartur) [inline snforge_std/src/cheatcodes.cairo:4]: These shouldn't be pub

File: snforge_std/src/cheatcodes/events.cairo [hunk 2]
```diff
@@ -15,20 +15,20 @@ pub fn spy_events() -> EventSpy {
...
+    pub keys: Array<felt252>,
+    pub data: Array<felt252>
 }
 
 /// An event spy structure allowing to get events emitted only after its creation.
 #[derive(Drop, Serde)]
-struct EventSpy {
-    _event_offset: usize
+pub struct EventSpy {
+    pub _event_offset: usize
```
Comment (cptartur) [inline snforge_std/src/cheatcodes/events.cairo:25]:
```suggestion
    event_offset: usize
```

File: snforge_std/src/cheatcodes/execution_info.cairo [hunk 3]
```diff
@@ -1,36 +1,36 @@
...
 #[derive(Serde, Drop, Copy)]
-struct CheatArguments<T> {
-    value: T,
-    span: CheatSpan,
-    target: ContractAddress,
+pub struct CheatArguments<T> {
+    pub value: T,
+    pub span: CheatSpan,
+    pub target: ContractAddress,
 }
```
Comment (cptartur) [inline snforge_std/src/cheatcodes/execution_info.cairo:30]:
```suggestion
struct CheatArguments<T> {
    value: T,
    span: CheatSpan,
    target: ContractAddress,
}
```

File: snforge_std/src/cheatcodes/execution_info.cairo [hunk 4]
```diff
@@ -1,36 +1,36 @@
...
 
 #[derive(Serde, Drop, Copy)]
-enum Operation<T> {
+pub enum Operation<T> {
```
Comment (cptartur) [inline snforge_std/src/cheatcodes/execution_info.cairo:33]:
```suggestion
enum Operation<T> {
```

File: snforge_std/src/cheatcodes/execution_info.cairo [hunk 5]
```diff
@@ -46,21 +46,21 @@ enum Operation<T> {
...
-    chain_id: Operation<felt252>,
-    nonce: Operation<felt252>,
+pub struct TxInfoMock {
+    pub version: Operation<felt252>,
+    pub account_contract_address: Operation<ContractAddress>,
+    pub max_fee: Operation<u128>,
+    pub signature: Operation<Span<felt252>>,
+    pub transaction_hash: Operation<felt252>,
+    pub chain_id: Operation<felt252>,
+    pub nonce: Operation<felt252>,
```
Comment (cptartur) [inline snforge_std/src/cheatcodes/execution_info.cairo:56]:
```suggestion
struct TxInfoMock {
    version: Operation<felt252>,
    account_contract_address: Operation<ContractAddress>,
    max_fee: Operation<u128>,
    signature: Operation<Span<felt252>>,
    transaction_hash: Operation<felt252>,
    chain_id: Operation<felt252>,
    nonce: Operation<felt252>,
```

File: snforge_std/src/cheatcodes/execution_info.cairo [hunk 6]
```diff
@@ -46,21 +46,21 @@ enum Operation<T> {
...
-    paymaster_data: Operation<Span<felt252>>,
-    nonce_data_availability_mode: Operation<u32>,
-    fee_data_availability_mode: Operation<u32>,
-    account_deployment_data: Operation<Span<felt252>>,
+    pub resource_bounds: Operation<Span<ResourcesBounds>>,
+    pub tip: Operation<u128>,
+    pub paymaster_data: Operation<Span<felt252>>,
+    pub nonce_data_availability_mode: Operation<u32>,
+    pub fee_data_availability_mode: Operation<u32>,
+    pub account_deployment_data: Operation<Span<felt252>>,
```
Comment (cptartur) [inline snforge_std/src/cheatcodes/execution_info.cairo:63]:
```suggestion
    resource_bounds: Operation<Span<ResourcesBounds>>,
    tip: Operation<u128>,
    paymaster_data: Operation<Span<felt252>>,
    nonce_data_availability_mode: Operation<u32>,
    fee_data_availability_mode: Operation<u32>,
    account_deployment_data: Operation<Span<felt252>>,
```

File: snforge_std/src/cheatcodes/execution_info.cairo [hunk 7]
```diff
@@ -93,13 +93,13 @@ impl TxInfoMockImpl of Default<TxInfoMock> {
...
 /// - `StopGlobal` - reset to the initial value for all contracts
 #[derive(Copy, Drop, Serde)]
-struct BlockInfoMock {
-    block_number: Operation<u64>,
-    block_timestamp: Operation<u64>,
-    sequencer_address: Operation<ContractAddress>,
+pub struct BlockInfoMock {
+    pub block_number: Operation<u64>,
+    pub block_timestamp: Operation<u64>,
+    pub sequencer_address: Operation<ContractAddress>,
```
Comment (cptartur) [inline snforge_std/src/cheatcodes/execution_info.cairo:99]:
```suggestion
struct BlockInfoMock {
    block_number: Operation<u64>,
    block_timestamp: Operation<u64>,
    sequencer_address: Operation<ContractAddress>,
```

File: snforge_std/src/cheatcodes/execution_info.cairo [hunk 8]
```diff
@@ -93,13 +93,13 @@ impl TxInfoMockImpl of Default<TxInfoMock> {
...
 }
 
-impl BlockInfoMockImpl of Default<BlockInfoMock> {
+pub impl BlockInfoMockImpl of Default<BlockInfoMock> {
```
Comment (cptartur) [inline snforge_std/src/cheatcodes/execution_info.cairo:102]:
```suggestion
impl BlockInfoMockImpl of Default<BlockInfoMock> {
```

File: snforge_std/src/cheatcodes/execution_info.cairo [hunk 9]
```diff
@@ -119,13 +119,13 @@ impl BlockInfoMockImpl of Default<BlockInfoMock> {
...
 /// - `StopGlobal` - reset to the initial value for all contracts
 #[derive(Copy, Drop, Serde)]
-struct ExecutionInfoMock {
-    block_info: BlockInfoMock,
-    tx_info: TxInfoMock,
-    caller_address: Operation<ContractAddress>,
+pub struct ExecutionInfoMock {
+    pub block_info: BlockInfoMock,
+    pub tx_info: TxInfoMock,
+    pub caller_address: Operation<ContractAddress>,
```
Comment (cptartur) [inline snforge_std/src/cheatcodes/execution_info.cairo:125]:
```suggestion
struct ExecutionInfoMock {
    block_info: BlockInfoMock,
    tx_info: TxInfoMock,
    caller_address: Operation<ContractAddress>,
```

File: snforge_std/src/cheatcodes/execution_info.cairo [hunk 10]
```diff
@@ -119,13 +119,13 @@ impl BlockInfoMockImpl of Default<BlockInfoMock> {
...
 }
 
-impl ExecutionInfoMockImpl of Default<ExecutionInfoMock> {
+pub impl ExecutionInfoMockImpl of Default<ExecutionInfoMock> {
```
Comment (cptartur) [inline snforge_std/src/cheatcodes/execution_info.cairo:128]:
```suggestion
impl ExecutionInfoMockImpl of Default<ExecutionInfoMock> {
```

File: snforge_std/src/cheatcodes/execution_info.cairo [hunk 11]
```diff
@@ -140,7 +140,7 @@ impl ExecutionInfoMockImpl of Default<ExecutionInfoMock> {
 /// Changes `ExecutionInfo` returned by `get_execution_info()`
 /// - `execution_info_mock` - a struct with same structure as `ExecutionInfo` (returned by
 /// `get_execution_info()`)
-fn cheat_execution_info(execution_info_mock: ExecutionInfoMock) {
+pub fn cheat_execution_info(execution_info_mock: ExecutionInfoMock) {
```
Comment (cptartur) [inline snforge_std/src/cheatcodes/execution_info.cairo:143]:
```suggestion
fn cheat_execution_info(execution_info_mock: ExecutionInfoMock) {
```

File: snforge_std/src/cheatcodes/l1_handler.cairo [hunk 12]
```diff
@@ -4,12 +4,12 @@ use starknet::{ContractAddress, testing::cheatcode, SyscallResult};
 use super::super::_cheatcode::handle_cheatcode;
 
 #[derive(Drop, Clone)]
-struct L1Handler {
-    target: ContractAddress,
-    selector: felt252,
+pub struct L1Handler {
+    pub target: ContractAddress,
+    pub selector: felt252,
 }
```
Comment (cptartur) [inline snforge_std/src/cheatcodes/l1_handler.cairo:10]:
```suggestion
    target: ContractAddress,
    selector: felt252,
}
```

File: snforge_std/src/cheatcodes/message_to_l1.cairo [hunk 13]
```diff
@@ -14,26 +14,26 @@ fn spy_messages_to_l1() -> MessageToL1Spy {
...
-    payload: Array<felt252>
+    pub payload: Array<felt252>
 }
 
 /// A message spy structure allowing to get messages emitted only after its creation.
 #[derive(Drop, Serde)]
-struct MessageToL1Spy {
-    _message_offset: usize
+pub struct MessageToL1Spy {
+    pub _message_offset: usize
```
Comment (cptartur) [inline snforge_std/src/cheatcodes/message_to_l1.cairo:27]:
```suggestion
    message_offset: usize
```

File: snforge_std/src/cheatcodes/storage.cairo [hunk 14]
```diff
@@ -2,12 +2,14 @@ use core::array::ArrayTrait;
...
 use core::option::OptionTrait;
 use core::traits::TryInto;
-use starknet::{testing::cheatcode, ContractAddress, storage_address_try_from_felt252};
+use starknet::{testing::cheatcode, ContractAddress, StorageAddress};
 use core::panic_with_felt252;
 use super::super::_cheatcode::handle_cheatcode;
 
 fn validate_storage_address_felt(storage_address_felt: felt252) {
-    match storage_address_try_from_felt252(storage_address_felt) {
+    let result: Option<StorageAddress> = storage_address_felt.try_into();
```
Comment (cptartur) [inline snforge_std/src/cheatcodes/storage.cairo:10]: Note for the future: this is caused by this signature `pub(crate) extern fn storage_address_try_from_felt252`

File: snforge_std/src/fs/file_operations.cairo [hunk 15]
```diff
@@ -4,11 +4,11 @@ use super::super::byte_array::byte_array_as_felt_array;
 use super::super::_cheatcode::handle_cheatcode;
 
 #[derive(Drop, Clone)]
-struct File {
-    path: ByteArray
+pub struct File {
+    pub path: ByteArray
```
Comment (cptartur) [inline snforge_std/src/fs/file_operations.cairo:8]:
```suggestion
    path: ByteArray
```

File: snforge_std/src/_cheatcode.cairo
Comment (cptartur) [inline snforge_std/src/_cheatcode.cairo:1]: `handle_cheatcode` should be `pub(crate)`.

**Review by cptartur — CHANGES_REQUESTED** [review]

File: snforge_std/src/cheatcodes/execution_info.cairo (hunk 4 above)
Comment (kkawula) [reply snforge_std/src/cheatcodes/execution_info.cairo:33]: I suppose that it should stay public:
https://github.com/foundry-rs/starknet-foundry/blob/master/crates/forge/tests/integration/cheat_execution_info.rs#L151-L155

It also shows that `CheatArguments` should be public

File: crates/forge/tests/integration/runtime.rs [hunk 16]
```diff
@@ -59,34 +59,3 @@ fn cairo_test_cheatcode_error() {
...
-        indoc!(
-            r#"
-                "Reading from buffer failed, this can be caused by calling starknet::testing::cheatcode with invalid arguments.
-                    Probably `snforge_std`/`sncast_std` version is incompatible, check above for incompatibility warning.
-                    "
-            "#
-        ),
-    );
-    assert_failed(&result);
-}
```
Comment (cptartur) [inline crates/forge/tests/integration/runtime.rs:92]: Do we have an issue to re-add these tests somewhere in cairo?

File: crates/forge/tests/integration/cheat_execution_info.rs [hunk 17]
```diff
@@ -100,178 +100,6 @@ fn start_and_stop_cheat_transaction_hash_single_attribute() {
     assert_passed(&result);
 }
 
-#[test]
-#[allow(clippy::too_many_lines)]
-fn start_cheat_execution_info_all_attributes_mocked() {
-    let test = test_case!(
```
Comment (cptartur) [inline crates/forge/tests/integration/cheat_execution_info.rs:106]: Let's `#[ignore]` these removed tests instead and add a TODO comment with relevant issue linked

**Review by cptartur — CHANGES_REQUESTED** [review]

**Review by Draggu — CHANGES_REQUESTED** [review]
I see that after making parts of it private, it is hard to test, but please create an issue to research ways of testing private parts of `snforge_std`. Also, keep these tests with `#[ignore]` instead of removing them entirely.

File: crates/forge/tests/integration/test_state.rs [hunk 18]
```diff
@@ -585,7 +584,6 @@ fn spy_events_simple() {
             fn spy_events_simple() {
                 let contract_address = test_address();
                 let mut spy = spy_events();
-                assert(spy._event_offset == 0, 'Events offset should be 0');
```
Comment (Draggu) [inline crates/forge/tests/integration/test_state.rs:588]: add TODO here with issue num

File: snforge_std/src/_cheatcode.cairo [hunk 19]
```diff
@@ -18,7 +18,8 @@ pub fn handle_cheatcode(input: Span<felt252>) -> Span<felt252> {
     }
 }
 
-fn _is_config_run() -> bool {
+// Do not use this function directly.
```
Comment (Draggu) [inline snforge_std/src/_cheatcode.cairo:21]: Please describe that it is an internal part of the snforge architecture used by macros. `Do not use this function directly.` raises a question: "Why?".

File: snforge_std/src/lib.cairo [hunk 20]
```diff
@@ -1,148 +1,142 @@
...
+
+pub mod env;
+
+pub mod signature;
+
+pub mod trace;
+
+pub mod byte_array;
+
+pub mod _cheatcode;
```
Comment (Draggu) [inline snforge_std/src/lib.cairo:140]: I think we should not expose the whole module. Re-export needed part only (`_is_config_run()`).

File: crates/forge/tests/integration/runtime.rs (hunk 16 above)
Comment (kkawula) [reply crates/forge/tests/integration/runtime.rs]: https://github.com/foundry-rs/starknet-foundry/issues/2765

File: crates/forge/tests/integration/cheat_execution_info.rs [hunk 21]
```diff
@@ -117,7 +118,6 @@ fn start_cheat_execution_info_all_attributes_mocked() {
             use array::SpanTrait;
             use snforge_std::{ declare, ContractClassTrait, DeclareResultTrait, cheat_execution_info, ExecutionInfoMock, Operation, CheatArguments, CheatSpan};
             use starknet::info::v2::ResourceBounds;
-
```
Comment (Draggu) [inline crates/forge/tests/integration/cheat_execution_info.rs:120]: keep these lines, formatting should not be a part of this PR

File: crates/forge/tests/integration/message_to_l1.rs [hunk 22]
```diff
@@ -494,7 +494,6 @@ fn test_filtering() {
                 let second_address = second_dispatcher.contract_address;
 
                 let mut spy = spy_messages_to_l1();
-                assert(spy._message_offset == 0, 'Message offset should be 0');
```
Comment (Draggu) [inline crates/forge/tests/integration/message_to_l1.rs:497]: TODO

File: crates/forge/tests/integration/spy_events.rs [hunk 23]
```diff
@@ -49,7 +49,6 @@ fn spy_events_simple() {
                 let dispatcher = ISpyEventsCheckerDispatcher { contract_address };
 
                 let mut spy = spy_events();
-                assert(spy._event_offset == 0, 'Event offset should be 0');
```
Comment (Draggu) [inline crates/forge/tests/integration/spy_events.rs:52]: TODO

File: crates/forge/tests/integration/spy_events.rs [hunk 24]
```diff
@@ -779,7 +778,6 @@ fn test_filtering() {
                 let second_dispatcher = ISpyEventsCheckerDispatcher { contract_address: second_address };
 
                 let mut spy = spy_events();
-                assert(spy._event_offset == 0, 'Event offset should be 0');
```
Comment (Draggu) [inline crates/forge/tests/integration/spy_events.rs:782]: TODO

**Review by Draggu — APPROVED** [review]

**Pull request merged**